### DISCUSSION TITLE
How to login an already registered user directly when they visit a protected resource using a token in the query params.

### DISCUSSION THREAD

**Discussion by green-mint** [supabase/supabase#16878] [Questions]
Hi there, 

I hope this is not a dumb question, I could not find a simpler/builtin method to do this.

### What I am trying to do.

I am using Nextjs 13 (app router) with the supabase-next-helpers. My application sends an email to a user with the url to a protected resource. When they click the url, if they are not signed it, they get redirected to the auth page. 

I want the users to directly login (if they are not already logged in) when they click the url to the protected resource.

One way that this could be done is to add a unique token to the query params in the url of the protected resource. And when the url is clicked, I can check the unique token in the middleware, fetch the user's latest access token and refresh tokens from the database and use the `supabase.auth.setSession` function to login the user.

Is there any better or more efficient method to achieve what I am trying to do?

**Comment by vladimirzb** [comment]
Hi! I'm relatively new to Supabase and Next.js 13, but I'm handling a similar scenario in my project using middleware. The main difference is that I use `auth.getUser` instead of `setSession`. Below is a code snippet that shows how I'm doing it. I hope this helps!

**Context**: This is a middleware function that should be placed in your `middleware` directory.

```jsx
import { createMiddlewareClient } from '@supabase/auth-helpers-nextjs'
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'
import { Database } from './types/supabase'

export async function middleware(req: NextRequest) {
  const res = NextResponse.next()
  const supabase = createMiddlewareClient<Database>({ req, res })

  const {
    data: { user },
  } = await supabase.auth.getUser()

  // Redirect signed-in users to /ingresos if they are on the root path
  if (user && req.nextUrl.pathname === '/') {
    return NextResponse.redirect(new URL('/ingresos', req.url))
  }

  // Redirect users who are not signed in to the root path if they are not already there
  if (!user && req.nextUrl.pathname !== '/') {
    return NextResponse.redirect(new URL('/', req.url))
  }

  return res
}

export const config = {
  matcher: ['/', '/account'],
}
```

**Note**: My solution doesn't handle the token in the query params, as you mentioned in your question. You might need to add that part to this middleware function to fully meet your requirements.
